### PR TITLE
Automatically close response stream once client connection closes

### PR DESCRIPTION
**Pull request by clue**
Builds on top of #187
This is also a prerequisite for upcoming changes related to #98

### CODE REVIEW

File: src/Server.php [hunk 1]
```diff
@@ -339,15 +339,37 @@ private function handleResponseBody(ResponseInterface $response, ConnectionInter
...
+            if ($response->getHeaderLine('Transfer-Encoding') === 'chunked') {
+                $stream = new ChunkedEncoder($stream);
+            }
+
+            // Close response stream once connection closes.
+            // Note that this TCP/IP close detection may take some time,
+            // in particular this may only fire on a later read/write attempt
+            // because we stop/pause reading from the connection once the
+            // request has been processed.
+            $connection->on('close', function() use ($stream) {
```
Comment (WyriHaximus) [inline src/Server.php]: Might as well write this as `$connection->on('close', [$stream, 'close']);`?

Comment (clue) [inline src/Server.php]: I have updated the PR accordingly (have no strong opinion here) :+1:

File: tests/FunctionalServerTest.php [hunk 2]
```diff
@@ -351,6 +352,100 @@ public function testSecureHttpsOnHttpStandardPortReturnsUriWithPort()
...
+
+        $response = Block\await($result, $loop, 1.0);
+
+        $this->assertStringStartsWith("HTTP/1.0 200 OK", $response);
+        $this->assertStringEndsWith("\r\n\r\n", $response);
+
+        $socket->close();
+    }
+
+    public function testStreamFromRequestHandlerWillBeClosedIfConnectionClosesWhileSendingBody()
```
Comment (WyriHaximus) [inline tests/FunctionalServerTest.php]: Test fails: https://travis-ci.org/reactphp/http/jobs/235196355#L223

File: tests/FunctionalServerTest.php [hunk 3]
```diff
@@ -351,6 +352,100 @@ public function testSecureHttpsOnHttpStandardPortReturnsUriWithPort()
...
+
+        $response = Block\await($result, $loop, 1.0);
+
+        $this->assertStringStartsWith("HTTP/1.0 200 OK", $response);
+        $this->assertStringEndsWith("\r\n\r\n", $response);
+
+        $socket->close();
+    }
+
+    public function testStreamFromRequestHandlerWillBeClosedIfConnectionClosesButWillOnlyBeDetectedOnNextWrite()
```
Comment (WyriHaximus) [inline tests/FunctionalServerTest.php]: This test also fails: https://travis-ci.org/reactphp/http/jobs/235196355#L228

File: tests/FunctionalServerTest.php (hunk 2 above)
Comment (clue) [inline tests/FunctionalServerTest.php]: Fixed tests by increasing test timeouts :shipit:

File: tests/FunctionalServerTest.php (hunk 3 above)
Comment (clue) [inline tests/FunctionalServerTest.php]: Fixed tests by increasing test timeouts :shipit:

**Pull request merged**